File: app/src/main/java/br/com/valenstech/letraviva/viewmodel/DevocionalViewModel.java
package br.com.valenstech.letraviva.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import br.com.valenstech.letraviva.R;
import br.com.valenstech.letraviva.model.ConteudoDevocional;
import br.com.valenstech.letraviva.repository.CallbackResultado;
import br.com.valenstech.letraviva.repository.RepositorioDevocional;
import br.com.valenstech.letraviva.util.EstadoUi;

public class DevocionalViewModel extends AndroidViewModel {

    @NonNull
    private final RepositorioDevocional repositorioDevocional;
    @NonNull
    private final MutableLiveData<EstadoUi<ConteudoDevocional>> estado;

    public DevocionalViewModel(@NonNull Application aplicacao) {
        super(aplicacao);
        repositorioDevocional = new RepositorioDevocional();
        estado = new MutableLiveData<>();
    }

    @NonNull
    public LiveData<EstadoUi<ConteudoDevocional>> getEstado() {
        return estado;
    }

    public void carregarDevocional() {
        estado.setValue(new EstadoUi.Carregando<>());
        repositorioDevocional.buscarDevocionalMaisRecente(new CallbackResultado<ConteudoDevocional>() {
            @Override
            public void aoSucesso(ConteudoDevocional resultado) {
                if (resultado == null) {
                    String mensagem = getApplication().getString(R.string.empty_devotional_message);
                    estado.postValue(new EstadoUi.Vazio<>(mensagem));
                } else {
                    estado.postValue(new EstadoUi.Sucesso<>(resultado));
                }
            }

            @Override
            public void aoErro(@NonNull Exception excecao) {
                String mensagemErro = excecao.getMessage();
                if (mensagemErro == null || mensagemErro.trim().isEmpty()) {
                    mensagemErro = getApplication().getString(R.string.error_loading_devotional);
                }
                estado.postValue(new EstadoUi.Erro<>(mensagemErro));
            }
        });
    }
}

File: app/src/main/java/br/com/valenstech/letraviva/viewmodel/DevocionalViewModel.kt
package br.com.valenstech.letraviva.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import br.com.valenstech.letraviva.R
import br.com.valenstech.letraviva.model.DevotionalContent
import br.com.valenstech.letraviva.repository.DevotionalRepository
import br.com.valenstech.letraviva.util.UiState

class DevocionalViewModel(application: Application) : AndroidViewModel(application) {

    private val repository = DevotionalRepository()
    private val _state = MutableLiveData<UiState<DevotionalContent>>()
    val state: LiveData<UiState<DevotionalContent>> = _state

    fun loadDevotional() {
        _state.value = UiState.Loading
        repository.fetchLatestDevotional { result ->
            result.onSuccess { content ->
                if (content == null) {
                    _state.postValue(
                        UiState.Empty(
                            getApplication<Application>().getString(R.string.empty_devotional_message)
                        )
                    )
                } else {
                    _state.postValue(UiState.Success(content))
                }
            }.onFailure { exception ->
                _state.postValue(
                    UiState.Error(
                        exception.message
                            ?: getApplication<Application>().getString(R.string.error_loading_devotional)
                    )
                )
            }
        }
    }
}
